src/mscorlib/shared/System/IO/BinaryWriter.cs

```diff
@@ -389,6 +390,62 @@ public unsafe virtual void Write(String value)
             }
         }
 
+        public virtual void Write(ReadOnlySpan<byte> value)
+        {
+            if (GetType() == typeof(BinaryWriter))
+            {
+                OutStream.Write(value);
+            }
+            else
+            {
+                int length = value.Length;
+                byte[] buffer = ArrayPool<byte>.Shared.Rent(length);
+                try
+                {
+                    value.CopyTo(buffer);
+                    Write(buffer, 0, length);
+                }
+                finally
+                {
+                    ArrayPool<byte>.Shared.Return(buffer);
+                }
+            }
+        }
+
+        public virtual void Write(ReadOnlySpan<char> value)
+        {
+            if (GetType() == typeof(BinaryWriter))
+            {
+                int length = value.Length;
+                char[] chars = ArrayPool<char>.Shared.Rent(length);
+                try
+                {
+                    value.CopyTo(chars);
+                    byte[] bytes = _encoding.GetBytes(chars, 0, length);
+                    OutStream.Write(bytes);
+                }
+                finally
+                {
+                    ArrayPool<char>.Shared.Return(chars);
+                }
+            }
+            else
+            {
+                int length = value.Length;
+                char[] chars = ArrayPool<char>.Shared.Rent(length);
+                try
+                {
+                    value.CopyTo(chars);
+                    byte[] bytes = _encoding.GetBytes(chars, 0, length);
```

Comment:
We can get the bytes array from the pool and just use the span based encoding overloads, e.g.
```C#
public virtual void Write(ReadOnlySpan<char> value)
{
    byte[] bytes = ArrayPool<byte>.Shared.Rent(_encoding.GetMaxByteCount(value.Length));
    try
    {
        int bytesWritten = _encoding.GetBytes(value, bytes);
        Write(bytes, 0, bytesWritten);
    }
    finally
    {
        ArrayPool<byte>.Shared.Return(bytes);
    }
}
```